### PR TITLE
DEP: Deprecate np.sum(generator)

### LINKED ISSUE #10652: DOC/BUG: np.sum supports generators, unlike the other reduction helpers
```python
>>> np.sum(a for a in range(10))
45
>>> np.prod(a for a in range(10))
<generator object <genexpr> at 0x0000023BAF7ED1A8>
```

Generators are not considered `array_like`, so the result for `sum` does not match the documentation.

Introduced in a13aad3ac33b629f3e696b4d4d5dbf4b5605d567

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #10652

Introduced in a13aad3ac33b629f3e696b4d4d5dbf4b5605d567

(milestoned only to indicate that the release notes are 1.15)

I think the original rationale for this change was to allow the following change to not break this code
```diff
+from numpy import *

 def my_func(n):
     return sum(i*i for i in range(n))
```

Reasons to deprecate it:
* `sum` is the only function to do this, even though `any` and `all` also shadow builtins
* It's not documented
* It's not consistent with `np.array(generator).sum()`, which already fails.

### CODE REVIEW

**Comment by charris** [comment]
Well, lets give this a shot. It is old behavior, so someone somewhere is probably using it, OTOH, it is new enough that it shouldn't affect Konrad :)

Thanks Eric.

**Comment by rgommers** [comment]
Comment if we come back to this later: the warning message is misleading. This should error in the future, not just give a different result.

**Comment by eric-wieser** [comment]
@rgommers: I don't agree. Every other reduction ufunc works without error:
```python
>>> np.prod(i for i in range(10))
<generator object <genexpr> at 0x0000016843E16150>
>>> np.max(i for i in range(10))
<generator object <genexpr> at 0x0000016844A3D6D0>
>>> np.add.reduce(i for i in range(10))
<generator object <genexpr> at 0x0000016843E16150>
>>> np.sum(i for i in range(10))
45
```

If we change sum to throw an error, then we have to change `ufunc.reduce` too, which is almost impossible - the issue here is the implicit conversion from a generator to an 0d object array, which is an entirely separate issue (although clearly a valid one)

**Comment by rgommers** [comment]
Okay, let me rephrase. I don't have too much of an opinion on this change either way, but the deprecation message is not great and needs fixing. Everyone who sees that is left wondering what will happen next. `np.sum(i for i in range(10))` --> 45 makes sense, if the answer "changes", the user is left wondering whether it will be 46 perhaps in the next release ...

In this case returning the generator object itself is likely to lead to exceptions rather than changed numerical results for the user, so fairly safe.

> If we change sum to throw an error, then we have to change ufunc.reduce too

just a general comment here: there is *a lot* of churn at the moment to fix what are basically longstanding minor inconsistencies. Please don't think "have to" - sometimes it's okay to leave the inconsistencies in. I am missing a bit of a balanced view of both the cost of that churn, and whether or not all that work could be better spent on new features rather than polishing existing ones.

**Comment by eric-wieser** [comment]
> In this case returning the generator object itself is likely to lead to exceptions rather than changed numerical results for the user, so fairly safe.

I'd hope so, yes. I'm not sure how to describe this though - the error won't come from `sum` itself, but from the downstream user code. There seem to be three classes of change:

* Returns a numerically different result
* Returns a result of a different type / dtype - likely leading to an exception downstream somewhere
* Raises an exception directly

I think I've previously considered  the first two in the same bucket, but in this case you're right -that doing so just makes the warning seem more alarming than it is.

**Comment by rgommers** [comment]
> I'm not sure how to describe this though - the error won't come from sum itself, but from the downstream user code. 

How about something like "instead of returning the sum over the values that the generator yields, an object array of size (1,) with the generator in it will be returned"?

> Returns a result of a different type / dtype - likely leading to an exception downstream somewhere

It depends what dtype I'd say. E.g. changing float64 to float32 or complex to float would be pretty bad. This case is probably >98% certain to lead to an exception later on, so not nearly as problematic.

**Comment by OverLordGoldDragon** [comment]
`from numpy import *` is fairly edge-case for ridding of convenience of feeding generators, and is a discouraged practice anyway. Wouldn't it be better to introduce a kwarg to control this behavior? e.g. `allowgen=False` (default True), and throw a warning once per session when a generator is fed, or somehow detect the star import.
